crypto/crmf/crmf_lib.c

```diff
@@ -0,0 +1,769 @@
+    }
+    return (int)res;
+}
+
+int OSSL_CRMF_MSG_get_certReqId(OSSL_CRMF_MSG *crm)
+{
+    if (crm == NULL || /* not really needed: */ crm->certReq == NULL) {
+        CRMFerr(CRMF_F_OSSL_CRMF_MSG_GET_CERTREQID, CRMF_R_NULL_ARGUMENT);
+        return -1;
+    }
+    return CRMF_ASN1_get_int(CRMF_F_OSSL_CRMF_MSG_GET_CERTREQID,
+                             crm->certReq->certReqId);
+}
+
+
+int OSSL_CRMF_MSG_set0_extensions(OSSL_CRMF_MSG *crm,
+                                  X509_EXTENSIONS *exts)
+{
+    OSSL_CRMF_CERTTEMPLATE *tmpl;
+    if (crm == NULL) {
+        CRMFerr(CRMF_F_OSSL_CRMF_MSG_SET0_EXTENSIONS, CRMF_R_NULL_ARGUMENT);
+        return 0;
+    }
+    if ((tmpl = OSSL_CRMF_MSG_get0_tmpl(crm)) == NULL) {
+        CRMFerr(CRMF_F_OSSL_CRMF_MSG_SET0_EXTENSIONS, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+
+    if (sk_X509_EXTENSION_num(exts) <= 0) {
+        sk_X509_EXTENSION_free(exts);
+        exts = NULL; /* do not include empty extensions list */
+    }
+
+    tmpl->extensions = exts;
+    return 1;
+}
+
+
+int OSSL_CRMF_MSG_push0_extension(OSSL_CRMF_MSG *crm,
+                                  const X509_EXTENSION *ext)
+{
+    int new = 0;
+    OSSL_CRMF_CERTTEMPLATE *tmpl = OSSL_CRMF_MSG_get0_tmpl(crm);
+
+    if (tmpl == NULL || ext == NULL) {
+        CRMFerr(CRMF_F_OSSL_CRMF_MSG_PUSH0_EXTENSION, CRMF_R_NULL_ARGUMENT);
+        return 0;
+    }
+
+    if ((tmpl->extensions) == NULL) {
+        if ((tmpl->extensions = sk_X509_EXTENSION_new_null()) == NULL)
+            goto oom;
+        new = 1;
+    }
+
+    if (!sk_X509_EXTENSION_push(tmpl->extensions, (X509_EXTENSION *)ext))
+        goto oom;
+    return 1;
+ oom:
+    CRMFerr(CRMF_F_OSSL_CRMF_MSG_PUSH0_EXTENSION, ERR_R_MALLOC_FAILURE);
+
+    if (new != 0) {
+        sk_X509_EXTENSION_free(tmpl->extensions);
+        tmpl->extensions = NULL;
+    }
+    return 0;
+}
+
+/*
+ * TODO: also support cases 1+2 defined in RFC4211, section 4.1.
+ * returns pointer to created OSSL_CRMF_POPOSIGNINGKEY on success, NULL on error
+ */
+static OSSL_CRMF_POPOSIGNINGKEY *CRMF_poposigkey_new(OSSL_CRMF_CERTREQUEST *cr,
+                                                     const EVP_PKEY *pkey,
+                                                     int dgst)
+{
+    OSSL_CRMF_POPOSIGNINGKEY *ps = NULL;
+    int l;
+    size_t crlen, max_sig_size;
+    unsigned int siglen;
+    unsigned char *crder = NULL, *sig = NULL;
+    int alg_nid=0, md_nid=0;
+    const EVP_MD *alg = NULL;
+
+    EVP_MD_CTX *ctx = NULL;
+
+    if (cr == NULL || pkey == NULL) {
+        CRMFerr(CRMF_F_CRMF_POPOSIGKEY_NEW, CRMF_R_NULL_ARGUMENT);
+        return 0;
+    }
+
+    if ((ps = OSSL_CRMF_POPOSIGNINGKEY_new()) == NULL)
+        goto err;
+
+    /* OpenSSL defaults all bit strings to be encoded as ASN.1 NamedBitList */
+    ps->signature->flags &= ~(ASN1_STRING_FLAG_BITS_LEFT | 0x07);
+    ps->signature->flags |= ASN1_STRING_FLAG_BITS_LEFT;
+
+    l = i2d_OSSL_CRMF_CERTREQUEST(cr, &crder);
```

Comment:
I don't like this identifier, please don't use lower-case L, because I can't see the difference to 1.

Reply:
Yes; changed to 'len'.